### PR TITLE
Enforce even more the code blocks attributes check through rustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
`rustdoc` now has a lint which allows it to warn if a code block attribute is malformated (which can end up in bad situations, even more in case of testing examples!). Now it'll fail if such a situation is encountered when testing markdown code blocks examples.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Fine by me!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by Mark-Simulacrum** [comment]
Does rustdoc not support the rustc syntax of `-D<lint name>`? If it does, I would prefer to use that for consistency.

I would also like to move the deny in builder.rs to somewhere in here: https://github.com/rust-lang/rust/blob/2f517ce6f28b5d638cce4c1eccdbe63255b11420/src/bootstrap/builder.rs#L1145 and adjust the one in rustdoc_cmd to also use deny_warnings to enable deny.

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by Mark-Simulacrum** [comment]
Hm, I think the lint name has been updated since then to be plural but r=me with that fixed.

**Comment by GuillaumeGomez** [comment]
Ah the PR got merged already? Updating then!

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Any idea what's going in:

```
error[E0602]: unknown lint: `invalid_codeblock_attributes`
  |
  = help: did you mean: `invalid_codeblock_attribute`
  = note: requested on the command line with `-D invalid_codeblock_attributes`
```

? I tested locally and the option works just as expected. Is it using an older rustc/rustdoc version?

**Comment by Mark-Simulacrum** [comment]
Oh, yeah, right -- we're checking that `--stage 0 doc` works on CI. I think probably the easiest thing is just to wait ... 3 days or so until we bump the beta compiler anyway?

**Comment by GuillaumeGomez** [comment]
Let's wait for the new beta then.

File: src/bootstrap/builder.rs [hunk 1]
```diff
@@ -725,7 +725,11 @@ impl<'a> Builder<'a> {
             .env("CFG_RELEASE_CHANNEL", &self.config.channel)
             .env("RUSTDOC_REAL", self.rustdoc(compiler))
             .env("RUSTDOC_CRATE_VERSION", self.rust_version())
-            .env("RUSTC_BOOTSTRAP", "1");
+            .env("RUSTC_BOOTSTRAP", "1")
+            .arg("-Dinvalid_codeblock_attributes");
+        if self.config.deny_warnings {
+            cmd.arg("-Dwarnings");
+        }
```
Comment (ollie27) [inline src/bootstrap/builder.rs:732]: rustdoc ignores lints when run on markdown files so this has no effect.

File: src/bootstrap/builder.rs [hunk 2]
```diff
@@ -1146,6 +1150,7 @@ impl<'a> Builder<'a> {
             if mode == Mode::Std && compiler.stage == 0 && self.config.incremental {
                 rustflags.arg("-Aunused-attributes");
             }
+            rustdocflags.arg("-Dinvalid_codeblock_attributes");
```
Comment (ollie27) [inline src/bootstrap/builder.rs:1153]: `invalid_codeblock_attributes` is a warn by default lint so there's no reason to pass this. The `-Dwarnings` above already makes it deny.

File: src/librustdoc/core.rs [hunk 3]
```diff
@@ -315,7 +315,7 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
     let missing_doc_example = rustc_lint::builtin::MISSING_DOC_CODE_EXAMPLES.name;
     let private_doc_tests = rustc_lint::builtin::PRIVATE_DOC_TESTS.name;
     let no_crate_level_docs = rustc_lint::builtin::MISSING_CRATE_LEVEL_DOCS.name;
-    let invalid_codeblock_attribute_name = rustc_lint::builtin::INVALID_CODEBLOCK_ATTRIBUTES.name;
+    let invalid_codeblock_attributes_name = rustc_lint::builtin::INVALID_CODEBLOCK_ATTRIBUTES.name;
```
Comment (ollie27) [inline src/librustdoc/core.rs:318]: Good catch!

**Review by ollie27 — CHANGES_REQUESTED** [review]

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1153]: No, above is `rustflags` and we're adding it to `rustdocflags`. Or are you referring to another than the one 10 lines above?

Comment (ollie27) [reply src/bootstrap/builder.rs:1153]: Yeah, I was mistaken. We should really just start adding `-Dwarnings` to `rustdocflags` as well as `rustflags`.

Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1153]: That would be problematic because of the broken links for the moment.

File: src/bootstrap/builder.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:732]: This function isn't run only on markdown files. ;)

File: src/bootstrap/builder.rs [hunk 4]
```diff
@@ -725,7 +725,11 @@ impl<'a> Builder<'a> {
             .env("CFG_RELEASE_CHANNEL", &self.config.channel)
             .env("RUSTDOC_REAL", self.rustdoc(compiler))
             .env("RUSTDOC_CRATE_VERSION", self.rust_version())
-            .env("RUSTC_BOOTSTRAP", "1");
+            .env("RUSTC_BOOTSTRAP", "1")
+            .arg("-Dinvalid_codeblock_attributes");
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs:729]: This shouldn't be `-D...` but rather `-W`, with `-Dwarnings` upgrading that only when `--deny=warnings` or equivalent is passed to x.py.

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs:1165]: This should only be `-D` if `deny_warnings` is on.

**Comment by GuillaumeGomez** [comment]
Still not working with the new lint. :'(

File: src/bootstrap/builder.rs (hunk 1 above)
Comment (ollie27) [reply src/bootstrap/builder.rs:732]: What else is it used for? I can only see `rustdoc_cmd` being used for markdown files.

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1165]: No because we want this check all the time.

File: src/bootstrap/builder.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:732]: No you're right, my bad. I confused the book from `src/doc` with libstd for whatever reason... I'll remove this chance.

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs:1165]: ...No? I don't see how that follows. It should be a warning only if the developer has not requested to deny warnings.

**Comment by Mark-Simulacrum** [comment]
Yeah, https://github.com/rust-lang/rust/pull/74395 will need to land first.

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1165]: This is all this PR is about: prevent malformed doc codeblocks attributes in the libstd and the error codes. If I let it as a warning, it'll never be fixed and the situation will remain the same.

Comment (ollie27) [reply src/bootstrap/builder.rs:1165]: > That would be problematic because of the broken links for the moment.

What broken links? There shouldn't be any and passing `-Dwarnings` will prevent new ones being added. I can understand not wanting to pass `-Dwarnings` for the compiler docs so if that's the concern then just denying `invalid_codeblock_attributes` for now is fine. We should definitely pass `-Dwarnings` when building at least the `std` docs at some point though.

Also setting `rustdocflags` here will override the setting from https://github.com/rust-lang/rust/blob/4cd0ee9343da86d9770bf0a514a682d240e0dce8/src/bootstrap/doc.rs#L530-L534 and https://github.com/rust-lang/rust/blob/4cd0ee9343da86d9770bf0a514a682d240e0dce8/src/bootstrap/doc.rs#L627

Those need to be changed to use the same mechanism for setting `rustdocflags` as is used here.

Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1165]: > What broken links?

The intra-doc ones. We have a few warnings when generating (std) docs. But yes, once intra-doc links are stabilized, I intend to change it to `-Dwarnings`.

I'll update the `rustdocflags`.

Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs:1165]: > This is all this PR is about: prevent malformed doc codeblocks attributes in the libstd and the error codes. If I let it as a warning, it'll never be fixed and the situation will remain the same.

This is confusing to me as an assertion. On CI, deny warnings is always enabled. Locally, when iterating, it can be useful to have warnings as just warnings. I imagine that applies to this one too.

File: src/bootstrap/builder.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:729]: Same answer as I provided [here](https://github.com/rust-lang/rust/pull/71670#discussion_r455874412).

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1165]: Oh, this is an important information. Then I guess it's acceptable.

File: src/bootstrap/builder.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:732]: Now it makes sense because `invalid_codeblock_attributes` is set to warning level by default.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum I'll trust you on this one: I set the lint to warning level by default. I really hope it is set to deny when running in CI. :)

File: src/bootstrap/builder.rs [hunk 5]
```diff
@@ -1162,8 +1163,8 @@ impl<'a> Builder<'a> {
             // fixed via better support from Cargo.
             cargo.env("RUSTC_LINT_FLAGS", lint_flags.join(" "));
 
-            rustdocflags.arg("-Dinvalid_codeblock_attributes");
         }
+        rustdocflags.arg("-Winvalid_codeblock_attributes");
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs:1167]: Why was this moved outside of the InTree block? We should still only specify this for in-tree crates.

Comment (GuillaumeGomez) [reply src/bootstrap/builder.rs:1167]: Woups, my bad.

**Comment by GuillaumeGomez** [comment]
I think it's now ready @Mark-Simulacrum. :)
